File: core/sudoku_solver.py
"""SudokuSolver — 백트래킹 솔버 및 유일해 검증."""
from typing import List, Optional, Tuple

Grid = List[List[int]]


class SudokuSolver:
    SIZE = 9
    BOX = 3

    @staticmethod
    def _find_empty(grid: Grid) -> Optional[Tuple[int, int]]:
        for r in range(9):
            for c in range(9):
                if grid[r][c] == 0:
                    return (r, c)
        return None

    @staticmethod
    def _can_place(grid: Grid, row: int, col: int, num: int) -> bool:
        for i in range(9):
            if grid[row][i] == num or grid[i][col] == num:
                return False
        br, bc = (row // 3) * 3, (col // 3) * 3
        for r in range(br, br + 3):
            for c in range(bc, bc + 3):
                if grid[r][c] == num:
                    return False
        return True

    @classmethod
    def solve(cls, grid: Grid) -> bool:
        """grid를 in-place로 해결. 가능하면 True."""
        empty = cls._find_empty(grid)
        if not empty:
            return True
        r, c = empty
        for n in range(1, 10):
            if cls._can_place(grid, r, c, n):
                grid[r][c] = n
                if cls.solve(grid):
                    return True
                grid[r][c] = 0
        return False

    @classmethod
    def count_solutions(cls, grid: Grid, cap: int = 2) -> int:
        """해의 개수를 최대 cap개까지 셈."""
        count = [0]

        def rec():
            if count[0] >= cap:
                return
            empty = cls._find_empty(grid)
            if not empty:
                count[0] += 1
                return
            r, c = empty
            for n in range(1, 10):
                if count[0] >= cap:
                    return
                if cls._can_place(grid, r, c, n):
                    grid[r][c] = n
                    rec()
                    grid[r][c] = 0

        rec()
        return count[0]

    @classmethod
    def has_unique_solution(cls, grid: Grid) -> bool:
        return cls.count_solutions([row[:] for row in grid], cap=2) == 1

    @classmethod
    def get_hint(cls, grid: Grid) -> Optional[Tuple[int, int, int]]:
        """비어있는 한 셀의 정답을 반환 (row, col, num). 없으면 None."""
        solved = [row[:] for row in grid]
        if not cls.solve(solved):
            return None
        for r in range(9):
            for c in range(9):
                if grid[r][c] == 0:
                    return (r, c, solved[r][c])
        return None

File: Flutter/lib/core/sudoku_solver.dart
/// 수도쿠 솔버 (백트래킹)
/// Python sudoku_solver.py → Dart 1:1 변환
class SudokuSolver {
  /// 빈 칸 찾기 (row, col) 반환. 없으면 null.
  static (int, int)? _findEmpty(List<List<int>> grid) {
    for (int r = 0; r < 9; r++) {
      for (int c = 0; c < 9; c++) {
        if (grid[r][c] == 0) return (r, c);
      }
    }
    return null;
  }

  /// (row, col)에 num을 놓을 수 있는지 검증
  static bool canPlace(List<List<int>> grid, int row, int col, int num) {
    for (int i = 0; i < 9; i++) {
      if (grid[row][i] == num || grid[i][col] == num) return false;
    }
    final br = (row ~/ 3) * 3;
    final bc = (col ~/ 3) * 3;
    for (int r = br; r < br + 3; r++) {
      for (int c = bc; c < bc + 3; c++) {
        if (grid[r][c] == num) return false;
      }
    }
    return true;
  }

  /// 인플레이스 백트래킹 풀기. 풀리면 true.
  static bool solve(List<List<int>> grid) {
    final empty = _findEmpty(grid);
    if (empty == null) return true;
    final (r, c) = empty;
    for (int n = 1; n <= 9; n++) {
      if (canPlace(grid, r, c, n)) {
        grid[r][c] = n;
        if (solve(grid)) return true;
        grid[r][c] = 0;
      }
    }
    return false;
  }

  /// 해의 개수를 최대 cap까지 셈
  static int countSolutions(List<List<int>> grid, {int cap = 2}) {
    int count = 0;
    void rec() {
      if (count >= cap) return;
      final empty = _findEmpty(grid);
      if (empty == null) {
        count++;
        return;
      }
      final (r, c) = empty;
      for (int n = 1; n <= 9; n++) {
        if (count >= cap) return;
        if (canPlace(grid, r, c, n)) {
          grid[r][c] = n;
          rec();
          grid[r][c] = 0;
        }
      }
    }

    rec();
    return count;
  }

  /// 유일해 여부 확인
  static bool hasUniqueSolution(List<List<int>> grid) {
    final copy = List.generate(9, (r) => List<int>.from(grid[r]));
    return countSolutions(copy, cap: 2) == 1;
  }

  /// 힌트 1개 반환: (row, col, answer). 풀 수 없으면 null.
  static (int, int, int)? getHint(List<List<int>> grid) {
    final solved = List.generate(9, (r) => List<int>.from(grid[r]));
    if (!solve(solved)) return null;
    for (int r = 0; r < 9; r++) {
      for (int c = 0; c < 9; c++) {
        if (grid[r][c] == 0) return (r, c, solved[r][c]);
      }
    }
    return null;
  }
}
